### PR TITLE
Adding Docstrings to Python Files

### PR DESCRIPTION
**Pull request by mrT23**
PR Type:
**Documentation**

___
PR Description:
**This pull request focuses on adding docstrings to various Python files in the project. The docstrings provide a brief description of the purpose of each function and class, and detail the input parameters and return types/values. This will improve code readability and maintainability, and provide better support for developers and users of the project.**

___
PR Main Files Walkthrough:
-`ai_handler.py`: Added docstrings to the AiHandler class, its constructor, and the chat_completion method.
-`git_patch_processing.py`: Added docstrings to the extend_patch, omit_deletion_hunks, handle_patch_deletions, and convert_to_hunks_with_lines_numbers functions.
-`pr_processing.py`: Added docstrings to the get_pr_diff, pr_generate_extended_diff, and pr_generate_compressed_diff functions.
-`token_handler.py`: Added docstrings to the TokenHandler class, its constructor, the _get_system_user_tokens method, and the count_tokens method.
-`utils.py`: Added docstrings to the convert_to_markdown, parse_code_suggestion, try_fix_json, fix_json_escape_char, convert_str_to_datetime, and load_large_diff functions.

### CODE REVIEW

**Comment by mrT23** [comment]
Preparing review...

File: pr_agent/algo/ai_handler.py [hunk 1]
```diff
@@ -9,7 +9,17 @@
...
+    and provides a method for performing chat completions using the OpenAI ChatCompletion API.
+    """
+
     def __init__(self):
+        """
+        Initializes the OpenAI API key and other settings from a configuration file.
+        Raises a ValueError if the OpenAI key is missing.
+        """
         try:
             openai.api_key = settings.openai.key
```
Comment (mrT23) [inline pr_agent/algo/ai_handler.py:24]: **Suggestion:** The class `AiHandler` is missing type hints for its methods. Adding type hints can improve code readability and help catch potential errors.
```suggestion
class AiHandler:
    def __init__(self) -> None:
        try:
            openai.api_key = settings.openai.key

    @retry(exceptions=(APIError, Timeout, TryAgain, AttributeError),
           tries=OPENAI_RETRIES, delay=2, backoff=2, jitter=(1, 3))
    async def chat_completion(self, model: str, temperature: float, system: str, user: str) -> Tuple[str, str]:
        try:
            response = await openai.ChatCompletion.acreate(
                            model=model,
```

**Comment by mrT23** [comment]
Preparing answer...

**Comment by mrT23** [comment]
Question: r a t e   t h i s   P R   f r o m   1   t o   1 0

Answer:
I would rate this PR as a 9. 

The PR is well-structured and the addition of docstrings significantly improves the readability and maintainability of the code. The docstrings are comprehensive, providing clear explanations of the purpose of each class, method, and function, as well as their arguments and return values. This will be very helpful for other developers who need to understand or modify this code in the future.

However, there are a few minor areas for improvement:

1. In `ai_handler.py`, the docstring for the `chat_completion` method mentions that it raises `TryAgain` twice. It would be better to mention each exception only once.

2. In `git_patch_processing.py`, the docstring for `omit_deletion_hunks` method could be improved by providing a brief explanation of what a 'hunk' is. This would make it easier for developers who are not familiar with this term to understand the purpose of the method.

3. In `pr_processing.py`, the docstring for `get_pr_diff` method could be improved by providing more information about what 'diff minimization techniques' are used.

4. In `token_handler.py`, the docstring for `_get_system_user_tokens` method could be improved by providing more information about what 'system and user strings' are.

5. In `utils.py`, the docstring for `try_fix_json` method could be improved by providing more information about what 'broken or incomplete JSON messages' are and how they are fixed.

6. The PR could benefit from a more descriptive title and a detailed description. This would provide context for the changes and make it easier for reviewers to understand the purpose of the PR.

Overall, the PR is of high quality and the additions are valuable. With a few minor improvements, it could be a perfect 10.

**Comment by mrT23** [comment]
Preparing answer...

**Comment by mrT23** [comment]
Preparing pr description...

**Review by hussam789 — APPROVED** [review]

**Pull request merged**